File: backend/app/models/ai_model.py
from sqlmodel import Field, Text, SQLModel

from app.models.base import BaseModel


class AiModelBase:
    supplier: int = Field(nullable=False)  # openai: 1, azure: 2, google: 3
    name: str = Field(max_length=255, nullable=False)
    base_model: str = Field(max_length=255, nullable=False)
    default_model: bool = Field(default=False, nullable=False)
    llm_type: str = Field(max_length=20, nullable=False,
                          default='chat')  # 'chat' or 'embedding'
    # Whether this model is a backup model
    backup_model: bool = Field(default=False, nullable=False)


class AiModelDetail(BaseModel, AiModelBase, table=True):
    __tablename__ = "ai_model"
    id: int = Field(default=None, primary_key=True)
    api_key: str | None = Field(nullable=True)
    api_domain: str = Field(nullable=False)
    # 1: openai-sdk, 2: vllm-sdk
    protocol: int = Field(nullable=False, default=1)
    config: str = Field(sa_type=Text())
    status: int = Field(nullable=False, default=1)

File: frontend/src/types/aiModel.ts
export enum Supplier {
    OpenAI = 1,
    Azure = 2,
    Google = 3,
}

export const SupplierLabels: Record<Supplier, string> = {
    [Supplier.OpenAI]: 'OpenAI',
    [Supplier.Azure]: 'Azure',
    [Supplier.Google]: 'Google',
}

export enum Protocol {
    OpenAISdk = 1,
    VllmSdk = 2,
}

export const ProtocolLabels: Record<Protocol, string> = {
    [Protocol.OpenAISdk]: 'OpenAI SDK',
    [Protocol.VllmSdk]: 'vLLM SDK',
}

export interface AiModelConfigItem {
    key: string
    val: unknown
    name: string
}

export interface AiModel {
    id: number
    name: string
    base_model: string
    supplier: Supplier
    protocol: Protocol
    api_key?: string
    api_domain: string
    default_model: boolean
    backup_model: boolean
    llm_type: string
    config?: string
    status: number
    created_at?: string
    updated_at?: string
}

export interface AiModelCreatePayload {
    name: string
    base_model: string
    supplier: number
    protocol: number
    api_domain: string
    api_key: string
    default_model: boolean
    backup_model: boolean
    llm_type: string
    config_list: AiModelConfigItem[]
}

export interface AiModelUpdatePayload extends AiModelCreatePayload {
    id: number
}
